test/test_tracer.cpp

```diff
@@ -1175,3 +1176,116 @@ TEST_CASE(
+      {"tracestate", "competitor=stuff,dd=o:Nebraska;s:1;ah:choo"}}, // origin is different
+     {{"traceparent", "00-00000000000000000000000000000030-000000000000002a-01"},
+      {"tracestate", "dd=s:2;o:Kansas;ah:choo,competitor=stuff"}}},
+
+    {__LINE__, "ignore interlopers",
+     {PropagationStyle::DATADOG, PropagationStyle::B3, PropagationStyle::W3C},
+     {PropagationStyle::W3C},
+     {{"x-datadog-trace-id", "48"}, {"x-datadog-parent-id", "64"},
+      {"x-datadog-origin", "Kansas"}, {"x-datadog-sampling-priority", "2"},
+      {"x-b3-traceid", "00000000000000000000000000000030"},
+      {"x-b3-parentspanid", "000000000000002a"},
+      {"x-b3-sampled", "0"}, // sampling is different
+      {"traceparent", "00-00000000000000000000000000000030-0000000000000040-01"},
+      {"tracestate", "competitor=stuff,dd=o:Nebraska;s:1;ah:choo"}},
+     {{"traceparent", "00-00000000000000000000000000000030-000000000000002a-01"},
+      {"tracestate", "dd=s:2;o:Kansas;ah:choo,competitor=stuff"}}},
+
+    {__LINE__, "don't take tracestate if trace ID doesn't match",
+     {PropagationStyle::DATADOG, PropagationStyle::W3C},
+     {PropagationStyle::W3C},
+     {{"x-datadog-trace-id", "48"}, {"x-datadog-parent-id", "64"},
+      {"x-datadog-origin", "Kansas"}, {"x-datadog-sampling-priority", "2"},
+      {"traceparent", "00-00000000000000000000000000000031-0000000000000040-01"},
+      {"tracestate", "competitor=stuff,dd=o:Nebraska;s:1;ah:choo"}},
+     {{"traceparent", "00-00000000000000000000000000000030-000000000000002a-01"},
+      {"tracestate", "dd=s:2;o:Kansas"}}},
+
+    {__LINE__, "don't take tracestate if W3C extraction isn't configured",
+     {PropagationStyle::DATADOG, PropagationStyle::B3},
+     {PropagationStyle::W3C},
+     {{"x-datadog-trace-id", "48"}, {"x-datadog-parent-id", "64"},
+      {"x-datadog-origin", "Kansas"}, {"x-datadog-sampling-priority", "2"},
+      {"traceparent", "00-00000000000000000000000000000030-0000000000000040-01"},
+      {"tracestate", "competitor=stuff,dd=o:Nebraska;s:1;ah:choo"}},
+     {{"traceparent", "00-00000000000000000000000000000030-000000000000002a-01"},
+      {"tracestate", "dd=s:2;o:Kansas"}}},
+  }));
+  // clang-format on
+
+  CAPTURE(test_case.line);
+  CAPTURE(test_case.description);
+  CAPTURE(to_json(test_case.extraction_styles));
+  CAPTURE(to_json(test_case.injection_styles));
+  CAPTURE(test_case.extracted_headers);
+  CAPTURE(test_case.expected_injected_headers);
+
+  TracerConfig config;
+  config.defaults.service = "testsvc";
+  config.extraction_styles = test_case.extraction_styles;
+  config.injection_styles = test_case.injection_styles;
+  config.logger = std::make_shared<NullLogger>();
+
+  auto finalized_config = finalize_config(config);
+  REQUIRE(finalized_config);
+  Tracer tracer{*finalized_config, std::make_shared<MockIDGenerator>()};
+
+  MockDictReader reader{test_case.extracted_headers};
+  auto span = tracer.extract_span(reader);
```

Comment:
The assumption "this code won't generate a trace ID" is not what is under test here.

Yours is a good idea, but that code won't compile due to the definitions of `Span` and `Expected`.

Here's the best that I could come up with:
```c++
  Tracer tracer{*finalized_config, std::make_shared<MockIDGenerator>()};

  MockDictReader reader{test_case.extracted_headers};
  MockDictWriter writer;
  REQUIRE_NOTHROW([&]() {
    auto span = tracer.extract_span(reader);
    REQUIRE(span);
    span->inject(writer);
  }());

  REQUIRE(writer.items == test_case.expected_injected_headers);
```

It bothers me that now we're asserting the non-throwness of three statements instead of the one expression `tracer.extract_span(reader)`.

Another way around this is to use a level of indirection:
```c++
Optional<Expected<Span>> maybe_span;
REQUIRE_NOTHROW(maybe_span = tracer.extract_span(reader));
REQUIRE(maybe_span.has_value());
REQUIRE(*maybe_span);
Span& span = **maybe_span;
```

There are probably other ways to put a `REQUIRE_NOTHROW` in there, but since that behavior is not what is under test, I'll leave the code as is.